tests/FSharp.Test.Utilities/Xunit/Attributes/DirectoryAttribute.fs

```diff
@@ -0,0 +1,67 @@
+﻿namespace FSharp.Test.Utilities.Xunit.Attributes
+
+open System
+open System.Collections.Generic
+open System.Diagnostics.CodeAnalysis
+open System.IO
+open System.Reflection
+open Xunit
+open Xunit.Extensions
+open Xunit.Sdk
+
+open FSharp.Test.Utilities
+open FSharp.Test.Utilities.Assert
+open FSharp.Test.Utilities.Compiler
+open FSharp.Test.Utilities.Utilities
+
+/// Attribute to use with Xunit's TheoryAttribute.
+/// Takes a directory, relative to current test suite's root.
+/// Returns a CompilationUnit with encapsulated source code, error baseline and IL baseline (if any).
+[<AttributeUsage(AttributeTargets.Method, AllowMultiple = true, Inherited = true)>]
+[<NoComparison; NoEquality>]
+type DirectoryAttribute(dir: string) =
+    inherit DataAttribute()
+    do
+        if String.IsNullOrWhiteSpace(dir) then
+            invalidArg "dir" "Directory cannot be null, empty or whitespace only."
+    let directory = dir
+
+    let readFileOrDefault (path: string) : string option =
+        match File.Exists(path) with
+            | true -> Some <| File.ReadAllText path
+            | _ -> None
+
+    let createCompilationUnit path fs =
+        let fsSource = (path ++ fs) |> File.ReadAllText
+        let bslFilePath = path ++ (fs + ".bsl")
+        let ilFilePath  = path ++ (fs + ".il")
+        let bslSource = readFileOrDefault bslFilePath
+        let ilSource = readFileOrDefault ilFilePath
+
+        { Source         = Text fsSource
+          Baseline       =
+                Some { SourceFilename = Some (path ++ fs)
+                       OutputBaseline = bslSource
+                       ILBaseline     = ilSource }
+          Options        = []
+          OutputType     = Library
+          SourceKind     = SourceKind.Fsx
+          Name           = Some fs
+          IgnoreWarnings = false
+          References     = [] } |> FS
+
+    override _.GetData(_: MethodInfo) =
+        let absolutePath = Path.GetFullPath(directory)
+
+        if not (Directory.Exists(absolutePath)) then
+            failwith (sprintf "Directory does not exist: \"%s\"." absolutePath)
+
+        let fsFiles = Directory.GetFiles(absolutePath, "*.fs") |> Array.map Path.GetFileName
```

Comment:
Can we have a mechanism, that allows us to specifically exclude files from the match, I'm thinking of perhaps tests that are known failures.  Or perhaps an opt-in for specific files to test.  I would probably want to select a single file to verify in the VS test explorer.

Does that make sense?  it does in my head but explaining is hard.

Reply:
Sure, makes sense, I can add an optional exclude property.
Thanks.

Reply:
And perhaps an include one as well?

Reply:
Yeah, that actually will be pretty flexible.

Reply:
> I would probably want to select a single file to verify in the VS test explorer.

Speaking of this one - you will be able without using Include/Exclude.
Since we're using "Theory" and "Directory" is a pretty much data provider for it, there will be a test case generated for each .fs file in directory, which you can run individually.